### PR TITLE
Generate `transform` as a final def for cats-tagless friendliness

### LINKED ISSUE #1587: Minor annoyance: can't derive cats-tagless typeclasses for `Alg[_[_]]`s
given the generated

```scala
type FooApi[F[_]] = smithy4s.kinds.FunctorAlgebra[FooApiGen, F]
```

and a call to e.g. `cats.tagless.FunctorK.derived[FooApi]`, you'll get this error:

```scala
no implicit values were found that match type cats.tagless.FunctorK[
  [G[_$7]] =>>
    smithy4s.Transformation.PartiallyApplied[
      foo.bar.FooApiGen[
        [I, E, O, SI, SO] =>> G[O]]
    ]
]
```

If the `transform` method is removed from the algebra, it's fine again.

Perhaps we should be generating an extension method (or nothing) instead of a member method? The implementation of `transform` is just `Transformation.of[FooApiApiGen[F]](this)`, after all.

### PR DESCRIPTION
**Pull request by kubukoz**
Closes #1587 

## PR Checklist (not all items are relevant to all PRs)

- [ ] Added unit-tests (for runtime code)
- [ ] Added bootstrapped code + smoke tests (when the rendering logic is modified)
- [ ] Added build-plugins integration tests (when reflection loading is required at codegen-time)
- [ ] Added alloy compliance tests (when simpleRestJson protocol behaviour is expanded/updated)
- [ ] Updated dynamic module to match generated-code behaviour
- [ ] Added documentation
- [x] Updated changelog

### CODE REVIEW

File: modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala [hunk 1]
```diff
@@ -51,7 +51,7 @@ private[codegen] object ModelLoader {
     val modelsInJars = deps.flatMap { file =>
       Using.resource(
         // Note: On JDK13+, the second parameter is redundant.
-        FileSystems.newFileSystem(file.toPath(), null)
+        FileSystems.newFileSystem(file.toPath(), null: ClassLoader)
```
Comment (kubukoz) [inline modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala:54]: **thought:** this was causing trouble for my sbt (@ghostbuster91 also reported seeing this):

```
[error] [codegen2_12] /Users/kubukoz/projects/smithy4s/modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala:54:21: ambiguous reference to overloaded definition,
[error] [codegen2_12] both method newFileSystem in class FileSystems of type (x$1: java.nio.file.Path, x$2: java.util.Map[String, _])java.nio.file.FileSystem
[error] [codegen2_12] and  method newFileSystem in class FileSystems of type (x$1: java.nio.file.Path, x$2: ClassLoader)java.nio.file.FileSystem
[error] [codegen2_12] match argument types (java.nio.file.Path,Null)
[error] [codegen2_12]         FileSystems.newFileSystem(file.toPath(), null)
[error] [codegen2_12]                     ^
[error] [codegen2_12] one error found
[error] [codegen2_12] (codegen2_12 / Compile / compileIncremental) Compilation failed
```

I'm not sure why it's failing now, but I'm sure it was fine a couple months ago. Anyway, doesn't seem important enough to investigate.

**Review by Baccata — APPROVED** [review]

File: modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala (hunk 1 above)
Comment (kubukoz) [reply modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala:54]: explanation for future generations: https://github.com/disneystreaming/smithy4s/pull/1580#discussion_r1817473402

tl;dr the "JRE only" distribution of temurin used in the Nix flake was missing the ct.sym file, meaning Scalac would see all JDK methods, ignoring its `-release` flag (the target JVM version) - so it was seeing JDK methods that it shouldn't have seen.
